### PR TITLE
ctrl: fix starting VM with Manual RunStrategy

### PR DESCRIPTION
**Pull request by acardace**
check for the returned error as the `vmi` variable being not nil does not necessarily mean that the VMI actually exists.

Because of this bug if you had a VM with `Manual` RunStrategy and you were to start the VM while the DataVolumes were not ready the just issued StartRequest would get trimmed by the virt-controller. To workaround this you'd have to issue a second start request when the DVs would get ready.

Unfortunately, our production and test code differ in this behavior as the production code always returns `vmi != nil` and the fake client returns the `vmi == nil` if the VMI does not exist.

### What this PR does
Before this PR:
 If you had a VM with `Manual` RunStrategy and you were to start the VM while the DataVolumes were not ready the just issued StartRequest would get trimmed by the virt-controller.

After this PR:
Start requests are never trimmed if the VMI doesn't exist.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
fix starting VM with Manual RunStrategy
```

### CODE REVIEW

**Comment by acardace** [comment]
@xpivarc @fossedihelm I think we need to backport this to v1.2 and cut a new minor release.

**Review by fossedihelm — APPROVED** [review]

**Comment by brianmcarey** [comment]
@acardace - looks like there are some related test failures here - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12016/pull-kubevirt-e2e-k8s-1.30-sig-compute/1796176218735775744

**Comment by acardace** [comment]
> @acardace - looks like there are some related test failures here - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12016/pull-kubevirt-e2e-k8s-1.30-sig-compute/1796176218735775744

Yup, they're fixed now.

**Comment by acardace** [comment]
@xpivarc @fossedihelm I'm pretty sure we're going to have to override the operator lane, it's unlikely that we'll see it passing before cutting a new release.

**Comment by dhiller** [comment]
Agree.

/override pull-kubevirt-e2e-k8s-1.30-sig-operator

File: pkg/virt-controller/watch/vm.go [hunk 1]
```diff
@@ -1295,6 +1295,10 @@ func (c *VMController) startVMI(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachi
 	log.Log.Object(vm).Infof("Started VM by creating the new virtual machine instance %s", vmi.Name)
 	c.recorder.Eventf(vm, k8score.EventTypeNormal, SuccessfulCreateVirtualMachineReason, "Started the virtual machine by creating the new virtual machine instance %v", vmi.ObjectMeta.Name)
 
+	if hasStartRequest(vm) {
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:1298]: Why is this required?

Comment (acardace) [reply pkg/virt-controller/watch/vm.go:1298]: Bunch of reasons, first off is simplification, why delaying something if you already know it can be done? 

Second, if you remove those pops() in Start()/Stop() the newly added unit test does not pass as `isTrimFirstChangeRequestNeeded()` only trims a Stop request if `vmi == nil` which is never the case if the VMI gets immediately deleted by the API server (like in unit tests), hence I smell somewhat a racy behaviour here. Basically the stop requests never get trimmed if deletion is very quick.

Instead of fixing the `isTrimFirstChangeRequestNeeded` I opted for this because I think it's just easier and more logical to act immediately, what I'd like to do next in a separate PR is removing that trim logic entirely.

File: pkg/virt-controller/watch/vm.go [hunk 2]
```diff
@@ -2820,8 +2819,6 @@ func (c *VMController) isTrimFirstChangeRequestNeeded(vm *virtv1.VirtualMachine,
 		// If we do not update `vmi` by asking the API Server this function could
 		// erroneously trim the just added StartRequest because it would see a running
 		// vmi with no DeletionTimestamp
-		vmi, _ := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Get(context.Background(), vm.GetName(), metav1.GetOptions{})
-		// If the current VMI is running, then it has been started.
 		if vmi != nil && vmi.DeletionTimestamp == nil {
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/vm.go:2822]: I know it is not related to these changes but i wonder why do we check if  `vmi.DeletionTimestamp == nil`

**Review by fossedihelm — APPROVED** [review]

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by xpivarc** [comment]
/hold 
just to give a chance to react to comments

File: pkg/virt-controller/watch/vm_test.go [hunk 3]
```diff
@@ -6556,6 +6556,52 @@ var _ = Describe("VirtualMachine", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				addVirtualMachine(vm)
+
+				sanityExecute(vm)
+
+				vm, err = virtFakeClient.KubevirtV1().VirtualMachines(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(vm).ToNot(BeNil())
+
+				Expect(vm.Status.StateChangeRequests).ToNot(BeEmpty())
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm_test.go:6601]: :+1:

File: pkg/virt-controller/watch/vm_test.go [hunk 4]
```diff
@@ -6556,6 +6556,52 @@ var _ = Describe("VirtualMachine", func() {
...
+				vm, err := virtFakeClient.KubevirtV1().VirtualMachines(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				addVirtualMachine(vm)
+
+				sanityExecute(vm)
+
+				_, err = virtFakeClient.KubevirtV1().VirtualMachineInstances(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+				Expect(err).To(MatchError(k8serrors.IsNotFound, "IsNotFound"))
+
+				By("Add Start Request")
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm_test.go:6588]: I guess to be sure.

File: pkg/virt-controller/watch/vm_test.go [hunk 5]
```diff
@@ -6641,6 +6687,69 @@ var _ = Describe("VirtualMachine", func() {
...
+				syncCache(controller.crIndexer, 1)
+
+				vmi, err := virtFakeClient.KubevirtV1().VirtualMachineInstances(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				// let the controller pick up the creation
+				vmiFeeder.Add(vmi)
+				sanityExecute(vm)
+
+				By("Add a stop and start Request")
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm_test.go:6712]: :+1:

File: pkg/virt-controller/watch/vm_test.go [hunk 6]
```diff
@@ -6641,6 +6687,69 @@ var _ = Describe("VirtualMachine", func() {
...
+				By("VM should get now restarted")
+				// pick up deletion
+				crSource.Delete(createVMRevision(vm))
+				syncCache(controller.crIndexer, 0)
+				vmiFeeder.Delete(vmi)
+				sanityExecute(vm)
+
+				// update VM from Store
+				vm, err = virtFakeClient.KubevirtV1().VirtualMachines(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm_test.go:6735]: :+1:

File: pkg/virt-controller/watch/vm.go [hunk 7]
```diff
@@ -2791,14 +2795,9 @@ func (c *VMController) isTrimFirstChangeRequestNeeded(vm *virtv1.VirtualMachine,
 	switch stateChange.Action {
 	case virtv1.StopRequest:
 		if vmi == nil {
-			// because either the VM or VMI informers can trigger processing here
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm.go:2794]: I'd rather leave it like this for consistency.

File: pkg/virt-controller/watch/vm_test.go [hunk 8]
```diff
@@ -6641,6 +6691,81 @@ var _ = Describe("VirtualMachine", func() {
...
+
+				// let the controller pick up the creation
+				vmiSource.Add(vmi)
+				addVirtualMachine(vm)
+				sanityExecute(vm)
+
+				// update VM from Store
+				vm, err = virtFakeClient.KubevirtV1().VirtualMachines(vm.Namespace).Get(context.TODO(), vm.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-controller/watch/vm_test.go]: Fixed it.

**Comment by xpivarc** [comment]
/hold
Unit test failure is relevant

**Comment by dhiller** [comment]
@acardace @xpivarc @fossedihelm I thought that this PR would also fix the clustered failures appearing in the operator lane. Am I mistaken there?

I see that https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12042/pull-kubevirt-e2e-k8s-1.30-sig-operator/1798281892731031552 which is based on the [merge commit containing this PR](https://github.com/kubevirt/kubevirt/commit/a7e0311d8704663351abd4bc9bbc8511753d2838) (as can be seen in the job yaml) is still showing those cluster failures.

Can you please help me understand this?

**Comment by fossedihelm** [comment]
@dhiller It will once the release 1.2.2 is published. The operator lane failures come from the release-1.2.1 and until we don't publish a new z release, they will not be fixed.

**Comment by dhiller** [comment]
> @dhiller It will once the release 1.2.2 is published. The operator lane failures come from the release-1.2.1 and until we don't publish a new z release, they will not be fixed.

@fossedihelm thank you for making me understand this :100: !
